### PR TITLE
fix(es/minifier): should not drop statements which has side-effects

### LINKED ISSUE #6422: Minifier doesn't respect side-effect in getter or setter
### Describe the bug

Minifier doesn't respect side-effect in getter or setter

### Input code

```typescript
let getter_effect = 'FAIL';
let setter_effect = 'FAIL';
let proto = {
	get foo() {
		getter_effect = 'PASS';
	},
	set bar(value) {
		setter_effect = 'PASS';
	}
};
let obj1 = {
	__proto__: proto
};
let obj2 = {
	__proto__: proto
};
let unused = obj1.foo;
obj2.bar = 0;

assert.strictEqual(getter_effect, 'PASS');
assert.strictEqual(setter_effect, 'PASS');
```

### Config

```json
{
  "jsc": {
    "transform": {
      "react": {
        "refresh": {}
      }
    },
    "parser": {
      "syntax": "ecmascript",
      "jsx": true
    },
    "target": "es2022",
    "loose": false,
    "minify": {
      "compress": {
        "arguments": false,
        "arrows": true,
        "booleans": true,
        "booleans_as_integers": false,
        "collapse_vars": true,
        "comparisons": true,
        "computed_props": true,
        "conditionals": true,
        "dead_code": true,
        "directives": true,
        "drop_console": false,
        "drop_debugger": true,
        "evaluate": true,
        "expression": false,
        "hoist_funs": false,
        "hoist_props": true,
        "hoist_vars": false,
        "if_return": true,
        "join_vars": true,
        "keep_classnames": false,
        "keep_fargs": true,
        "keep_fnames": false,
        "keep_infinity": false,
        "loops": true,
        "negate_iife": true,
        "properties": true,
        "reduce_funcs": false,
        "reduce_vars": false,
        "side_effects": true,
        "switches": true,
        "typeofs": true,
        "unsafe": false,
        "unsafe_arrows": false,
        "unsafe_comps": false,
        "unsafe_Function": false,
        "unsafe_math": false,
        "unsafe_symbols": false,
        "unsafe_methods": false,
        "unsafe_proto": false,
        "unsafe_regexp": false,
        "unsafe_undefined": false,
        "unused": true,
        "const_to_let": true,
        "pristine_globals": true
      },
      "mangle": false
    }
  },
  "module": {
    "type": "es6"
  },
  "isModule": true,
  "minify": false
}
```

### Playground link

https://play.swc.rs/?version=1.3.15&code=H4sIAAAAAAAAA32PsQrCMBCG5%2BQpsrWFUtTR4NBBQXAQ%2BgAhrVdRQqPJxUX67l6NiFLtmLvv%2Fu%2BPARRHQASnoG2hQbESyabc7hLJDe38xO7iLFqa3TmjCNFam2bDg40C92VV0RHrc84oUdTapTdtAkR%2BJHnzvI8qW5%2Fn0aTUU6vUMvo%2FiMU0Ebrg4UDMEFZQW8mHo4LK0HAmOdfeg8PCozs1uL4GbdKvr%2BSvZpn8hfo%2F6AO30%2BKDYwEAAA%3D%3D&config=H4sIAAAAAAAAA21TQbIiIQy9C2sXUy5m4QFm989AIYSWPw3pSoLfLsu7T7CVtnV25PHySF7C1XyzN4erEXKFI1JuAYHzshwiAZ%2F0eLvddmZyxEDtguci7mIOBnx27ClNYnaqpZBQBeWKowGkMXj%2Fa7%2FX2xGRwRyiGxl2JqeS4ty0POZJX%2BF21qSaoQh3niPCH15kd%2BaIOIJW%2Bh5bxzYVgQFoTfU4jm5isGdHPaO95igxlg1UBYKdCKcXtIQkCYuKPbEALliPATqQCLykM6wU1VBKYS2sV3IHAxzrMDT%2FFiac3ViddC243G3QF3veCROLjbXwG7SpdIGWJh%2BsFC2BVCpPzjemsvHhL4DWOTrm4jKsmXc86hy2zPgfVipRhyhzR3XCa1UFBm3OphR7h61qIEmrWwShemgd%2BlX8AW4a4hTAQozqdk%2FmnyT%2BtIrJPAHGHqprLq4zWEL73Kct2lbgA%2FyjVcnrOB54dnJ6x3jORxw%2FJDLICcMHrEYIvoOk23uZ3tFaAqjNEF4uKrewr6kOX9CO7bc9fdaF0Bw7jHjs66ufMrsy9LVsPzpjqA243s27%2F9bfRvHEX4%2BbRfH5W5fEf5vNM1k2BAAA

### Expected behavior

__Notice that terser also generated wrong code, the `obj1.foo` should not be removed.__

image

### Actual behavior

```js
let getter_effect = 'FAIL', setter_effect = 'FAIL';
assert.strictEqual(getter_effect, 'PASS'), assert.strictEqual(setter_effect, 'PASS');
```

### Version

1.3.15

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by iheyunfei**
**Description:**

**BREAKING CHANGE:**

**Related issue (if exists):**

Fix #6422

### CODE REVIEW

File: crates/swc_ecma_utils/src/lib.rs [hunk 1]
```diff
@@ -1375,18 +1375,29 @@ pub trait ExprExt {
...
-                        let is_static_accessor = |prop: &PropOrSpread| {
-                            if let PropOrSpread::Prop(prop) = prop {
-                                if let Prop::Getter(_) | Prop::Setter(_) = &**prop {
-                                    true
-                                } else {
-                                    false
-                                }
-                            } else {
-                                false
+                        let is_prop_could_has_side_effect = |prop: &PropOrSpread| {
```
Comment (kdy1) [inline crates/swc_ecma_utils/src/lib.rs:1378]: Can you rename this to `can_have_side_effect`?

Comment (iheyunfei) [reply crates/swc_ecma_utils/src/lib.rs]: Done.  Could you re-run the wasm test? I saw it failed due to 
> Run curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
sh: 139: [: x86_64: unexpected operator
info: downloading wasm-pack
info: successfully installed wasm-pack to `/home/runner/.cargo/bin/wasm-pack`
Segmentation fault (core dumped)
Error: Process completed with exit code 1.

File: crates/swc_ecma_utils/src/lib.rs [hunk 2]
```diff
@@ -1375,18 +1375,24 @@ pub trait ExprExt {
...
-                        let is_static_accessor = |prop: &PropOrSpread| {
-                            if let PropOrSpread::Prop(prop) = prop {
-                                if let Prop::Getter(_) | Prop::Setter(_) = &**prop {
-                                    true
-                                } else {
-                                    false
+                        let can_have_side_effect = |prop: &PropOrSpread| match prop {
+                            PropOrSpread::Spread(_) => true,
+                            PropOrSpread::Prop(prop) => match prop.as_ref() {
+                                Prop::Shorthand(i) => i.sym == js_word!("__proto__"),
```
Comment (kdy1) [inline crates/swc_ecma_utils/src/lib.rs:1381]: I thought about it, and I think it's right to have the same chord style

Comment (kdy1) [inline crates/swc_ecma_utils/src/lib.rs:1381]: Please change this to `Prop::Shorthand(Ident{ sym: js_word!("__proto__"),.. })`

File: crates/swc_ecma_utils/src/lib.rs [hunk 3]
```diff
@@ -1375,18 +1375,24 @@ pub trait ExprExt {
...
+                                Prop::KeyValue(KeyValueProp { key, .. }) => match key {
```
Comment (kdy1) [inline crates/swc_ecma_utils/src/lib.rs:1382]: Same here, you can merge it

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you! I'm leaving a bump comment to avoid forgetting correct bump.

---

swc-bump:
 - swc_atoms

**Review by jridgewell — APPROVED** [review]

**Review by kdy1 — APPROVED** [review]

**Pull request merged**